core/arch/arm/plat-zynq/main.c

```diff
@@ -0,0 +1,290 @@
+
+		/*
+		 * SCU Access Register : SAC = 0x0000000F
+		 * - both secure CPU access SCU
+		 */
+		write32(0x0000000F, SCU_BASE + SCU_SAC);
+
+		/*
+		 * SCU NonSecure Access Register : SNSAC : 0x00000FFF
+		 * - both nonsec cpu access SCU, private and global timer
+		 */
+		write32 (0x00000FFF, SCU_BASE + SCU_NSAC);
+
+		/* SCU enable */
+
+		write32(read32 (SCU_BASE + SCU_CTRL) | 0x1, SCU_BASE + SCU_CTRL);
+
+		/* NS Access control */
+
+		write32(0xFFFFFFFF, SECURITY2_SDIO0);
+		write32(0xFFFFFFFF, SECURITY3_SDIO1);
+		write32(0xFFFFFFFF, SECURITY4_QSPI);
+		write32(0xFFFFFFFF, SECURITY6_APB_SLAVES);
+
+		write32(SLCR_UNLOCK_MAGIC, SLCR_UNLOCK);
+
+		write32(0xFFFFFFFF, SLCR_TZ_DDR_RAM);
+		write32(0xFFFFFFFF, SLCR_TZ_DMA_NS);
+		write32(0xFFFFFFFF, SLCR_TZ_DMA_IRQ_NS);
+		write32(0xFFFFFFFF, SLCR_TZ_DMA_PERIPH_NS);
+		write32(0xFFFFFFFF, SLCR_TZ_GEM);
+		write32(0xFFFFFFFF, SLCR_TZ_SDIO);
+		write32(0xFFFFFFFF, SLCR_TZ_USB);
+
+		write32(SLCR_LOCK_MAGIC, SLCR_LOCK);
+	}
+	/*
+	 * GIC configuration
+	 *
+	 * per-CPU interrupts config are in accordance with GIC driver:
+	 *
+	 * ID0-ID7(SGI)   for Non-secure interrupts
+	 * ID8-ID15(SGI)  for Secure interrupts.
+	 * All PPI config as Non-secure interrupts.
+	 *
+	 * Register ICDISR0 = 0xFFFF00FF
+	 *
+	 * Register ICCPMR = 0x80
+	 */
+
+	write32(0xFFFF00FF, GIC_BASE + GICD_OFFSET + GIC_DIST_ISR0);
+	write32(0x80, GIC_BASE + GICC_OFFSET + CORE_ICC_ICCPMR);
+}
+
+static vaddr_t console_base(void)
+{
+	static void *console_base_va __data;
+
+	if (cpu_mmu_enabled()) {
+		if (!console_base_va)
+			console_base_va = phys_to_virt(CONSOLE_UART_BASE,
+					  MEM_AREA_IO_NSEC);
+		return (vaddr_t)console_base_va;
+	}
+	return CONSOLE_UART_BASE;
+}
+
+void console_init(void)
+{
+}
+
+void console_putc(int ch)
+{
+	if (ch == '\n')
+		cdns_uart_putc('\r', console_base());
+	cdns_uart_putc(ch, console_base());
+}
+
+void console_flush(void)
+{
+	cdns_uart_flush(console_base());
+}
+
+vaddr_t pl310_base(void)
+{
+	static void *va __data; /* in case it's used before .bss is cleared */
+
+	if (cpu_mmu_enabled()) {
+		if (!va)
+			va = phys_to_virt(PL310_BASE, MEM_AREA_IO_SEC);
+		return (vaddr_t)va;
+	}
+	return PL310_BASE;
+}
+
+void main_init_gic(void)
+{
+	vaddr_t gicc_base;
+	vaddr_t gicd_base;
+
+	gicc_base = (vaddr_t)phys_to_virt(GIC_BASE + GICC_OFFSET,
+					  MEM_AREA_IO_SEC);
+	gicd_base = (vaddr_t)phys_to_virt(GIC_BASE + GICD_OFFSET,
+					  MEM_AREA_IO_SEC);
+
+	if (!gicc_base || !gicd_base)
+		panic();
+
+	/* Initialize GIC */
+	gic_init(&gic_data, gicc_base, gicd_base);
+
+	itr_init(&gic_data.chip);
+}
+
+static vaddr_t slcr_access_range[] = {
+	0x004, 0x008,	/* lock, unlock */
+	0x100, 0x1FF,	/* PLL */
+	0x200, 0x2FF,	/* Reset */
+	0xA00, 0xAFF	/* L2C */
+};
+
+static uint32_t write_slcr(uint32_t addr, uint32_t val)
+{
+	uint32_t i;
+
+	for (i = 0; i < ARRAY_SIZE(slcr_access_range); i+=2) {
+		if (addr >= slcr_access_range[i] &&
+		addr <= slcr_access_range[i+1]) {
```

Comment:
should I align the second 'addr' at the same column of the first 'addr', and it means adding a few spaces instead of tab? Sorry I am not familiar with this.